### PR TITLE
Create a way to use a fireant.dataset.datasets as joins in dataset

### PR DESCRIPTION
**Pull request by x8lucas8x**
The main contribution in this is the DataSetBlender class, which mimics most of the DataSet API. That means you can use data_set_blender_instance.query... and data_set_blender_instance.latest... the same way. Also, the API of the DataSet class, in terms of fields and joins, was changed to use a builder pattern.

Usage example:

```
 ...

mock_dataset_blender = DataSetBlender(mock_dataset).join(mock_spend_dataset).field(
    'average-candidate-spend-per-voters',
    label='Candidate Spend per Voters',
    definition=mock_spend_dataset.fields['candidate-spend'] / mock_dataset.fields['voters'],
    data_type=DataType.number
).field(
    'average-candidate-spend-per-wins',
    label='Candidate Spend per Wins',
    definition=mock_spend_dataset.fields['candidate-spend'] / mock_dataset.fields['wins'],
    data_type=DataType.number
)
```

See tests for the aforementioned DataSetBlender on https://github.com/kayak/fireant/pull/253/files#diff-f4128f84d7aa2c2b06a3c8d8827a4069.

### CODE REVIEW

File: fireant/tests/queries/test_build_joins.py [hunk 1]
```diff
@@ -27,6 +30,46 @@ def test_dimension_with_join_includes_join_in_query(self):
                          'GROUP BY "$timestamp","$district-name" '
                          'ORDER BY "$timestamp","$district-name"', str(queries[0]))
 
+    def test_dimension_with_dataset_join_includes_a_sub_query_join_in_query(self):
```
Comment (x8lucas8x) [inline fireant/tests/queries/test_build_joins.py]: This test shows how everything is dynamically applied to both main and sub-queries. It includes filters too.

**Comment by x8lucas8x** [comment]
@twheys @RobinPapke I've added a lot more tests and fixed a couple corner cases. Please review the tests for the DataSetBlender on https://github.com/kayak/fireant/pull/253/files#diff-f4128f84d7aa2c2b06a3c8d8827a4069 once again.

File: fireant/dataset/data_blending.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        Adds a join when building a slicer query. The join type is left outer by default.
+
+        :param secondary_dataset: (Required)
+            The secondary dataset, which table will be used as part of the one or more JOIN expressions.
+        :param field_mapping: (Optional)
+            A dictionary with primary dataset's fields as keys and the provided secondary dataset's fields as values.
+            If not provided, fields with the same name will be matched by default.
+
+        :return:
+            A copy of the query with the join added.
```
Comment (RobinPapke) [inline fireant/dataset/data_blending.py]: The method returns an instance of the the DataSetBlender class with the join added

Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: Ah, yeah. I'll update this and the following method's docstring.

Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: @RobinPapke done

File: fireant/dataset/data_blending.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    DimensionChoicesQueryBuilder,
+    DimensionLatestQueryBuilder,
+)
+from fireant.queries.builder.dataset_blender_query_builder import DataSetBlenderQueryBuilder
+from fireant.utils import immutable
+
+
+class DataSetBlender(object):
+    """
+    WRITEME
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: Would be nice to have a brief description of the intended use of this class, given that it's somewhat high level

File: fireant/dataset/data_blending.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def __init__(self, primary_dataset):
+        """
+        Constructor for a data blended slicer.  Contains all the fields to initialize the slicer.
+
+        :param primary_dataset: (Required)
+            The primary dataset, which table will be used as part of the FROM expression.
+        """
+        self.primary_dataset = primary_dataset
+        self.table = primary_dataset.table
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: is there a benefit of storing a reference to this separately? What if, for example, someone using this class were to modify the primary dataset value, then this reference would become invalid, right?

File: fireant/dataset/data_blending.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        self.fields = DataSet.Fields()
+
+        self.field_mapping = {}
+
+        primary_dataset_field_mapping = {}
+
+        for primary_field in self.primary_dataset.fields:
+            primary_dataset_field_mapping[primary_field] = primary_field
+
+        self.field_mapping[primary_dataset.table._table_name] = primary_dataset_field_mapping
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: using the name name here is probably not a good idea. Perhaps `str(primary_dataset.table)` to get the fully qualified name and to avoid referencing private attributes?

File: fireant/dataset/data_blending.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            The primary dataset, which table will be used as part of the FROM expression.
+        """
+        self.primary_dataset = primary_dataset
+        self.table = primary_dataset.table
+        self.secondary_datasets = []
+        self.fields = DataSet.Fields()
+
+        self.field_mapping = {}
+
+        primary_dataset_field_mapping = {}
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: It's not immediately clear what this mapping is used for

File: fireant/dataset/data_blending.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+        :param primary_dataset: (Required)
+            The primary dataset, which table will be used as part of the FROM expression.
+        """
+        self.primary_dataset = primary_dataset
+        self.table = primary_dataset.table
+        self.secondary_datasets = []
+        self.fields = DataSet.Fields()
+
+        self.field_mapping = {}
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: not sure what this is used for either

File: fireant/dataset/data_blending.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        :return:
+            A copy of this DataSetBlender instance with the join added.
+        """
+        if field_mapping is None:
+            field_mapping = {}
+
+            for primary_field in self.primary_dataset.fields:
+                for secondary_field in secondary_dataset.fields:
+                    if primary_field.alias == secondary_field.alias:
+                        field_mapping[primary_field] = secondary_field
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: this nesting is really deep. Can you please consider refactoring it and simplifying the logic?

File: fireant/dataset/data_blending.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        ))
+
+    @immutable
+    def join(self, secondary_dataset, field_mapping=None):
+        """
+        Adds a join when building a slicer query. The join type is left outer by default.
+
+        :param secondary_dataset: (Required)
+            The secondary dataset, which table will be used as part of the one or more JOIN expressions.
+        :param field_mapping: (Optional)
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: would be really nice to provide some example of what this should look like in the docstring

File: fireant/dataset/data_blending.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+        :return:
+            A copy of this DataSetBlender instance with the field added.
+        """
+        field = Field(self, *args, **kwargs)
+
+        if not field.definition.is_aggregate:
+            field.choices = DimensionChoicesQueryBuilder(self.primary_dataset, field)
+
+        self.fields.append(
```
Comment (twheys) [inline fireant/dataset/data_blending.py]: minor style thing ... don't see any benefit of chopping down a single argument

File: fireant/queries/builder/dataset_blender_query_builder.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        self.all_datasets = [self.primary_dataset, *self.secondary_datasets]
+
+        if aggregation_func_mapping is None:
+            # SUM is used by default when no mapping is found
+            aggregation_func_mapping = {}
+
+        self.aggregation_func_mapping = aggregation_func_mapping
+
+    @property
+    def sql(self):
```
Comment (twheys) [inline fireant/queries/builder/dataset_blender_query_builder.py]: this is massive and needs a refactor. This could be a lot simpler if you would leverage the existing query builder code in the data set. Instead of repeating all of the logic, use the `DataSet.query` to build up a query in each of the data sets.

File: fireant/queries/builder/dataset_sub_query_builder.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
+from .query_builder import (
```
Comment (twheys) [inline fireant/queries/builder/dataset_sub_query_builder.py]: minor style thing ... could you make this single imports one line, please?

File: fireant/dataset/data_blending.py (hunk 7 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: This is for mapping fields across datasets. By default, if not provided, that matches by field names, but DataSetBlender users can override that too.

File: fireant/dataset/data_blending.py (hunk 4 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: I've made it a property.

File: fireant/dataset/data_blending.py (hunk 5 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: Done. I'm using the table instance now.

File: fireant/dataset/data_blending.py (hunk 7 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: I've added a comment here.

File: fireant/dataset/data_blending.py (hunk 8 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: I've refactored that.

File: fireant/dataset/data_blending.py (hunk 9 above)
Comment (x8lucas8x) [reply fireant/dataset/data_blending.py]: Examples were added.

File: fireant/queries/builder/dataset_sub_query_builder.py (hunk 12 above)
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_sub_query_builder.py]: Done.

File: fireant/queries/builder/dataset_blender_query_builder.py (hunk 11 above)
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder.py]: @twheys done the refactoring and reused the _DataSet.query_ to build up a query in each of the data sets. Using the sql method of the nested datasets didn't really made it simpler though, given most of the logic we need, either way, is related to the the sub query replacement, custom name changes in metrics/dimensions, deep copying of elements and corner cases (e.g. not applying dimension modifiers in the wrapping query, not applying dimension filters in the wrapping query, use actual aliases for referencing nested fields). I've added three extra tests for cases with references as well. Please see the tests for the DataSetBlender on https://github.com/kayak/fireant/pull/253/files#diff-f4128f84d7aa2c2b06a3c8d8827a4069 once again.

File: fireant/queries/builder/dataset_blender_query_builder/klass.py [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+
+            new_widgets = [copy.deepcopy(widget) for widget in self._widgets]
+            for widget in new_widgets:
+                widget.items = [copy.deepcopy(metric) for metric in dataset_metrics]
+                queries_per_dataset[dataset.table] = queries_per_dataset[dataset.table].widget(widget)
+
+        for dataset, dataset_filters in filters_per_dataset.items():
+            for filter in dataset_filters:
+                new_filter = copy.deepcopy(filter)
+                new_definition = copy.deepcopy(filter.definition)
```
Comment (twheys) [inline fireant/queries/builder/dataset_blender_query_builder/klass.py]: deepcopy should create a new instance of the definition, no? so this second deepcopy is redundant?

File: fireant/queries/builder/dataset_blender_query_builder/klass.py [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+        for metric in metrics:
+            for pypika_field in metric.definition.fields():
+                for dataset in self.all_datasets:
+                    try:
+                        metrics_per_dataset[dataset].append(
+                            dataset.fields[
+                                _get_pypika_field_name_or_alias(pypika_field).replace('_', '-').replace('$', '')
+                            ]
+                        )
+                    except AttributeError:
```
Comment (twheys) [inline fireant/queries/builder/dataset_blender_query_builder/klass.py]: could we put a comment to explain that these attributeerrors are meant to be ignored?

File: fireant/queries/builder/dataset_blender_query_builder/klass.py [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+    def __init__(self, dataset, aggregation_func_mapping=None):
+        super(DataSetBlenderQueryBuilder, self).__init__(dataset)
+        self.primary_dataset = dataset.primary_dataset
+        self.secondary_datasets = dataset.secondary_datasets
+
+        if aggregation_func_mapping is None:
+            # SUM is used by default when no mapping is found
+            aggregation_func_mapping = {}
+
+        self.aggregation_func_mapping = aggregation_func_mapping
```
[the remark this answers is not in the document]
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder/klass.py]: I meant to remove that. Going to do that now.

File: fireant/queries/builder/dataset_blender_query_builder/klass.py [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+    Blended slicer queries consist of widgets, dimensions, filters, orders by and references. At least one or
+    more widgets is required. All others are optional.
+    """
+
+    def __init__(self, dataset, aggregation_func_mapping=None):
+        super(DataSetBlenderQueryBuilder, self).__init__(dataset)
+        self.primary_dataset = dataset.primary_dataset
+        self.secondary_datasets = dataset.secondary_datasets
+
+        if aggregation_func_mapping is None:
```
[the remark this answers is not in the document]
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder/klass.py]: Replied in the comment below. This is going to be removed.

File: fireant/queries/builder/dataset_blender_query_builder/klass.py (hunk 13 above)
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder/klass.py]: I'll remove it.

File: fireant/queries/builder/dataset_blender_query_builder/klass.py [hunk 17]
```diff
@@ -77,40 +71,102 @@ def sql(self):
...
 
         sub_query_joins_groups, sub_query_tables_groups = self.__sub_query_groups
 
         queries = []
 
         for i, (sub_query_joins, sub_query_tables) in enumerate(zip(sub_query_joins_groups, sub_query_tables_groups)):
+            dimension_aliases = {alias_selector(el.alias) for el in self._dimensions}
+            metric_aliases = {alias_selector(el.alias) for el in metrics}
+            reference_aliases = {
+                alias_selector('{}_{}'.format(el, ref.alias))
```
[the remark this answers is not in the document]
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder/klass.py]: Sure. I'll change this to make use of it.

File: fireant/queries/builder/dataset_blender_query_builder/klass.py (hunk 14 above)
Comment (x8lucas8x) [reply fireant/queries/builder/dataset_blender_query_builder/klass.py]: Sure.
